src/engine/TransitivePathBinSearch.cpp

```diff
@@ -0,0 +1,245 @@
+ResultTable TransitivePathBinSearch::computeResult() {
+  if (minDist_ == 0 && !isBoundOrId() && lhs_.isVariable() &&
+      rhs_.isVariable()) {
+    AD_THROW(
+        "This query might have to evalute the empty path, which is currently "
+        "not supported");
+  }
+  auto [startSide, targetSide] = decideDirection();
+  shared_ptr<const ResultTable> subRes = subtree_->getResult();
+
+  IdTable idTable{allocator()};
+
+  idTable.setNumColumns(getResultWidth());
+
+  size_t subWidth = subRes->idTable().numColumns();
+
+  if (startSide.isBoundVariable()) {
+    shared_ptr<const ResultTable> sideRes =
+        startSide.treeAndCol_.value().first->getResult();
+    size_t sideWidth = sideRes->idTable().numColumns();
+
+    CALL_FIXED_SIZE((std::array{resultWidth_, subWidth, sideWidth}),
+                    &TransitivePathBinSearch::computeTransitivePathBound, this,
+                    &idTable, subRes->idTable(), startSide, targetSide,
+                    sideRes->idTable());
+
+    return {std::move(idTable), resultSortedOn(),
+            ResultTable::getSharedLocalVocabFromNonEmptyOf(*sideRes, *subRes)};
+  }
+  CALL_FIXED_SIZE((std::array{resultWidth_, subWidth}),
+                  &TransitivePathBinSearch::computeTransitivePath, this,
+                  &idTable, subRes->idTable(), startSide, targetSide);
+
+  // NOTE: The only place, where the input to a transitive path operation is not
+  // an index scan (which has an empty local vocabulary by default) is the
+  // `LocalVocabTest`. But it doesn't harm to propagate the local vocab here
+  // either.
+  return {std::move(idTable), resultSortedOn(), subRes->getSharedLocalVocab()};
+}
+
+// _____________________________________________________________________________
+TransitivePathBinSearch::Map TransitivePathBinSearch::transitiveHull(
+    const BinSearchMap& edges, const std::vector<Id>& startNodes,
+    std::optional<Id> target) const {
+  // For every node do a dfs on the graph
+  Map hull{allocator()};
+
+  std::vector<std::pair<Id, size_t>> stack;
+  ad_utility::HashSetWithMemoryLimit<Id> marks{
+      getExecutionContext()->getAllocator()};
+  for (auto startNode : startNodes) {
+    if (hull.contains(startNode)) {
+      // We have already computed the hull for this node
+      continue;
+    }
+
+    marks.clear();
+    stack.clear();
+    stack.push_back({startNode, 0});
+
+    if (minDist_ == 0 && (!target.has_value() || startNode == target.value())) {
+      insertIntoMap(hull, startNode, startNode);
+    }
+
+    while (stack.size() > 0) {
+      checkCancellation();
+      auto [node, steps] = stack.back();
+      stack.pop_back();
+
+      if (steps <= maxDist_ && marks.count(node) == 0) {
+        if (steps >= minDist_) {
+          marks.insert(node);
+          if (!target.has_value() || node == target.value()) {
+            insertIntoMap(hull, startNode, node);
+          }
+        }
+
+        auto successors = edges.successors(node);
```

Comment:
I think if you write a small wrapper, then this function is exactly the same as in the fallback and you can just use a template here
(maybe also write a hashMap wrapper that also has a `successors()` function etc.

Reply:
Is the current solution sufficient or should I implement the hashMap wrapper with a `successors()` function?

Reply:
I would really appreciate this, because then the two classes would be really almost the same, and the non-impl part would be really small.